src/test/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorServiceTest.java

```diff
@@ -24,22 +24,97 @@
 
 package org.jenkinsci.plugins.workflow.cps;
 
+import hudson.model.Result;
 import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
 import org.junit.ClassRule;
 import org.junit.Test;
 import org.junit.Rule;
+import org.junit.rules.ErrorCollector;
 import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
 import org.jvnet.hudson.test.JenkinsRule;
 
 public class CpsVmExecutorServiceTest {
 
     @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
     @Rule public JenkinsRule r = new JenkinsRule();
+    @Rule public ErrorCollector errors = new ErrorCollector();
 
     @Test public void contextClassLoader() throws Exception {
         WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
         p.setDefinition(new CpsFlowDefinition("echo(/yes I can load ${Thread.currentThread().contextClassLoader.loadClass(getClass().name)}/)", false));
         r.buildAndAssertSuccess(p);
     }
 
+    @Issue({"JENKINS-31314", "JENKINS-27306"})
+    @Test public void wrongCatcher() throws Exception {
+        boolean origFailOnMismatch = CpsVmExecutorService.FAIL_ON_MISMATCH;
+        CpsVmExecutorService.FAIL_ON_MISMATCH = false;
+        try {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("def ok() {sleep 1}; @NonCPS def bad() {for (int i = 0; i < 10; i++) {sleep 1}; assert false : 'never gets here'}; node {ok(); bad()}", true));
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "bad", null, "sleep"), r.buildAndAssertSuccess(p));
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("def l = [3, 2, 1]; println(/oops got ${l.sort {x, y -> x - y}}/)", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("oops got -1", b);
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("java.util.ArrayList", "sort", "org.jenkinsci.plugins.workflow.cps.CpsClosure2", "call"), b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("node {[1, 2, 3].each {x -> sleep 1; echo(/no problem got $x/)}}", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("no problem got 3", b);
+                r.assertLogNotContains("expected to call", b);
```

Comment:
When I test this PR locally against the latest version of the groovy-cps PR, this assertion fails, because the following line is present in the logs: `expected to call java.util.ArrayList.each but wound up catching ?`. Perhaps the build is passing here because the main snapshot is outdated? I am a little surprised that the expected receiver is `java.util.ArrayList.each` and not something like `$each__java_util_Iterable__groovy_lang_Closure`, since I thought this was one of the transformed methods from DefaultGroovyMethods.